sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableServiceImpl.java

```diff
@@ -210,6 +212,254 @@ public Row getCurrentRow() throws NoSuchElementException {
+      if (buffer.isEmpty()) {
+        return false;
+      }
+      currentRow = buffer.remove();
+      return currentRow != null;
+    }
+
+    Future<UpstreamResults> fetchNextSegment() {
+      SettableFuture<UpstreamResults> f = SettableFuture.create();
+      if (nextRequest == null) {
+        f.set(new UpstreamResults(ImmutableList.of(), null));
+        return f;
+      }
+
+      // TODO(diegomez): Remove atomic ScanHandler for simpler StreamObserver/Future implementation
+      AtomicReference<ScanHandler> atomic = new AtomicReference<>();
+      ScanHandler handler =
+          session
+              .getDataClient()
+              .readFlatRows(
+                  nextRequest,
+                  new StreamObserver<FlatRow>() {
+                    List<Row> rows = new ArrayList<>();
+                    long currentByteSize = 0;
+                    boolean byteLimitReached = false;
+
+                    @Override
+                    public void onNext(FlatRow flatRow) {
+                      currentByteSize += computeRowSize(flatRow);
+                      Row row = FlatRowConverter.convert(flatRow);
+                      rows.add(row);
+
+                      if (currentByteSize > bufferByteLimit) {
+                        byteLimitReached = true;
+                        atomic.get().cancel();
+                        return;
+                      }
+                    }
+
+                    @Override
+                    public void onError(Throwable e) {
+                      f.setException(e);
+                    }
+
+                    @Override
+                    public void onCompleted() {
+                      ReadRowsRequest nextNextRequest = null;
+
+                      if (byteLimitReached || rows.size() == segmentSize) {
+                        nextNextRequest =
+                            truncateRequest(nextRequest, rows.get(rows.size() - 1).getKey());
+                      }
+
+                      f.set(new UpstreamResults(rows, nextNextRequest));
+                    }
+                  });
+      atomic.set(handler);
+      return f;
+    }
+
+    void waitReadRowsFuture() throws IOException {
+      try {
+        UpstreamResults r = future.get();
+        buffer.addAll(r.rows);
+        nextRequest = r.nextRequest;
+        future = null;
+        serviceCallMetric.call("ok");
+      } catch (StatusRuntimeException e) {
+        serviceCallMetric.call(e.getStatus().getCode().value());
+        throw e;
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        throw new IOException(e);
+      } catch (ExecutionException e) {
+        throw new IOException(e);
+      }
+    }
+
+    ReadRowsRequest truncateRequest(ReadRowsRequest request, ByteString lastKey) {
+      RowSet setFromRequest = request.getRows();
+
+      RowSet.Builder segment = RowSet.newBuilder();
+      for (int i = 0; i < setFromRequest.getRowRangesCount(); i++) {
```

Comment:
why not a foreach of request.getRows()?

Reply:
Also this implementation doesn't care if the ranges are sorted...so either optimize it to or stop sorting the ranges in the beginning